File: sdk/typescript/src/budget.ts
/**
 * Per-user and global budget management.
 *
 * @module
 */

import { AgentCCError } from "./errors.js";

// ---------------------------------------------------------------------------
// Window parsing
// ---------------------------------------------------------------------------

const WINDOW_PATTERN = /^(\d+)([mhdwM])$/;

const WINDOW_UNIT_SECONDS: Record<string, number> = {
  m: 60,
  h: 3600,
  d: 86400,
  w: 604800,
  M: 2592000, // 30 days
};

function parseWindow(window: string): number {
  const match = WINDOW_PATTERN.exec(window);
  if (!match) {
    throw new Error(
      `Invalid window format: "${window}". ` +
        "Expected format like '1m', '5m', '1h', '1d', '1w', '1M'.",
    );
  }
  return parseInt(match[1], 10) * WINDOW_UNIT_SECONDS[match[2]];
}

// ---------------------------------------------------------------------------
// BudgetManager
// ---------------------------------------------------------------------------

export interface BudgetManagerOptions {
  maxBudget?: number;
  /** Rolling window duration, e.g. `"1h"`, `"1d"`, `"1M"`. */
  window?: string;
}

/**
 * Per-user and global budget management.
 *
 * Tracks cumulative spend and enforces budget limits at both global and
 * per-user levels.
 *
 * @example
 * ```typescript
 * const budget = new BudgetManager({ maxBudget: 10.0 }); // $10 total
 * budget.checkBudget(0.05);  // throws AgentCCError if over
 * budget.updateCost(0.05);
 *
 * // With rolling window:
 * const hourly = new BudgetManager({ maxBudget: 10.0, window: "1h" });
 * ```
 */
export class BudgetManager {
  readonly maxBudget: number | null;
  readonly window: string | null;

  private _currentSpend = 0;
  private _userBudgets: Record<string, number> = {};
  private _userSpend: Record<string, number> = {};
  private _windowStart: number;

  constructor(opts: BudgetManagerOptions = {}) {
    this.maxBudget = opts.maxBudget ?? null;
    this.window = opts.window ?? null;
    this._windowStart = Date.now() / 1000;
  }

  private _maybeResetWindow(): void {
    if (!this.window) return;
    const windowSeconds = parseWindow(this.window);
    const now = Date.now() / 1000;
    if (now - this._windowStart >= windowSeconds) {
      this._currentSpend = 0;
      this._userSpend = {};
      this._windowStart = now;
    }
  }

  /** Throws `AgentCCError` if adding `estimatedCost` would exceed budget. */
  checkBudget(estimatedCost: number = 0, user?: string): void {
    this._maybeResetWindow();
    if (
      this.maxBudget != null &&
      this._currentSpend + estimatedCost > this.maxBudget
    ) {
      throw new AgentCCError(
        `Budget exceeded: current spend $${this._currentSpend.toFixed(4)} + ` +
          `estimated $${estimatedCost.toFixed(4)} > max $${this.maxBudget.toFixed(4)}`,
      );
    }
    if (user && user in this._userBudgets) {
      const userSpend = this._userSpend[user] ?? 0;
      if (userSpend + estimatedCost > this._userBudgets[user]) {
        throw new AgentCCError(
          `User budget exceeded for '${user}': $${userSpend.toFixed(4)} + ` +
            `$${estimatedCost.toFixed(4)} > $${this._userBudgets[user].toFixed(4)}`,
        );
      }
    }
  }

  /** Record cost from a completed request. */
  updateCost(cost: number, user?: string): void {
    this._maybeResetWindow();
    this._currentSpend += cost;
    if (user) {
      this._userSpend[user] = (this._userSpend[user] ?? 0) + cost;
    }
  }

  /** Set budget limit for a specific user. */
  setUserBudget(user: string, budget: number): void {
    this._userBudgets[user] = budget;
  }

  /** Get current cumulative spend. */
  getCurrentSpend(user?: string): number {
    this._maybeResetWindow();
    if (user) return this._userSpend[user] ?? 0;
    return this._currentSpend;
  }

  /** Get remaining budget. Returns `null` if no budget is set. */
  getRemainingBudget(user?: string): number | null {
    this._maybeResetWindow();
    if (user && user in this._userBudgets) {
      return this._userBudgets[user] - (this._userSpend[user] ?? 0);
    }
    if (this.maxBudget != null) {
      return this.maxBudget - this._currentSpend;
    }
    return null;
  }

  /** Reset spend counters. */
  reset(user?: string): void {
    if (user) {
      this._userSpend[user] = 0;
    } else {
      this._currentSpend = 0;
      this._userSpend = {};
      this._windowStart = Date.now() / 1000;
    }
  }

  /**
   * Estimate future spend based on current rate.
   * @param hours - Number of hours to project (default 24).
   */
  projectedCost(hours: number = 24, user?: string): number {
    this._maybeResetWindow();
    const now = Date.now() / 1000;
    const elapsedHours = (now - this._windowStart) / 3600;
    if (elapsedHours <= 0) return 0;

    const current = user
      ? this._userSpend[user] ?? 0
      : this._currentSpend;
    return (current / elapsedHours) * hours;
  }

  /**
   * Check if a user has a budget set and hasn't exceeded it.
   */
  isValidUser(user: string): boolean {
    this._maybeResetWindow();
    if (!(user in this._userBudgets)) return false;
    const spend = this._userSpend[user] ?? 0;
    return spend <= this._userBudgets[user];
  }
}

File: sdk/python/src/agentcc/_budget.py
"""Per-user and global budget management."""

from __future__ import annotations

import re
import threading
import time
from dataclasses import dataclass, field

_WINDOW_PATTERN = re.compile(r"^(\d+)([mhdwM])$")

_WINDOW_UNIT_SECONDS: dict[str, float] = {
    "m": 60.0,
    "h": 3600.0,
    "d": 86400.0,
    "w": 604800.0,
    "M": 2592000.0,  # 30 days
}


@dataclass
class BudgetManager:
    """Per-user and global budget management.

    Tracks cumulative spend and enforces budget limits at both global and
    per-user levels.  Thread-safe via an internal lock.

    Usage::

        from agentcc import BudgetManager

        budget = BudgetManager(max_budget=10.0)  # $10 total
        budget.check_budget(0.05)  # raises AgentCCError if over
        budget.update_cost(0.05)

        # With rolling window:
        budget = BudgetManager(max_budget=10.0, window="1h")  # $10 per hour
    """

    max_budget: float | None = None
    window: str | None = None
    _current_spend: float = field(default=0.0, init=False)
    _user_budgets: dict[str, float] = field(default_factory=dict, init=False)
    _user_spend: dict[str, float] = field(default_factory=dict, init=False)
    _lock: threading.Lock = field(default_factory=threading.Lock, init=False)
    _window_start: float = field(default=0.0, init=False)

    def __post_init__(self) -> None:
        self._window_start = time.time()

    @staticmethod
    def _parse_window(window: str) -> float:
        """Convert a duration string (e.g. '1m', '5m', '1h', '1d', '1w', '1M') to seconds."""
        match = _WINDOW_PATTERN.match(window)
        if not match:
            raise ValueError(
                f"Invalid window format: {window!r}. "
                "Expected format like '1m', '5m', '1h', '1d', '1w', '1M'."
            )
        count = int(match.group(1))
        unit = match.group(2)
        return count * _WINDOW_UNIT_SECONDS[unit]

    def _maybe_reset_window(self) -> None:
        """Reset spend if the current window has expired. Must be called under lock."""
        if self.window is None:
            return
        window_seconds = self._parse_window(self.window)
        now = time.time()
        if now - self._window_start >= window_seconds:
            self._current_spend = 0.0
            self._user_spend.clear()
            self._window_start = now

    def check_budget(self, estimated_cost: float = 0.0, user: str | None = None) -> None:
        """Raise ``AgentCCError`` if adding *estimated_cost* would exceed budget."""
        with self._lock:
            self._maybe_reset_window()
            if self.max_budget is not None and self._current_spend + estimated_cost > self.max_budget:
                from agentcc._exceptions import AgentCCError

                raise AgentCCError(
                    f"Budget exceeded: current spend ${self._current_spend:.4f} + "
                    f"estimated ${estimated_cost:.4f} > max ${self.max_budget:.4f}"
                )
            if user and user in self._user_budgets:
                user_spend = self._user_spend.get(user, 0.0)
                if user_spend + estimated_cost > self._user_budgets[user]:
                    from agentcc._exceptions import AgentCCError

                    raise AgentCCError(
                        f"User budget exceeded for '{user}': ${user_spend:.4f} + "
                        f"${estimated_cost:.4f} > ${self._user_budgets[user]:.4f}"
                    )

    def update_cost(self, cost: float, user: str | None = None) -> None:
        """Record cost from a completed request."""
        with self._lock:
            self._maybe_reset_window()
            self._current_spend += cost
            if user:
                self._user_spend[user] = self._user_spend.get(user, 0.0) + cost

    def set_user_budget(self, user: str, budget: float) -> None:
        """Set budget limit for a specific user."""
        with self._lock:
            self._user_budgets[user] = budget

    def get_current_spend(self, user: str | None = None) -> float:
        """Get current cumulative spend."""
        with self._lock:
            self._maybe_reset_window()
            if user:
                return self._user_spend.get(user, 0.0)
            return self._current_spend

    def get_remaining_budget(self, user: str | None = None) -> float | None:
        """Get remaining budget.  Returns ``None`` if no budget is set."""
        with self._lock:
            self._maybe_reset_window()
            if user and user in self._user_budgets:
                return self._user_budgets[user] - self._user_spend.get(user, 0.0)
            if self.max_budget is not None:
                return self.max_budget - self._current_spend
            return None

    def reset(self, user: str | None = None) -> None:
        """Reset spend counters."""
        with self._lock:
            if user:
                self._user_spend[user] = 0.0
            else:
                self._current_spend = 0.0
                self._user_spend.clear()
                self._window_start = time.time()

    def projected_cost(self, hours: float = 24.0, user: str | None = None) -> float:
        """Estimate future spend based on current rate.

        Args:
            hours: Number of hours to project into the future (default 24).
            user: If provided, project for a specific user.

        Returns:
            Projected cost in USD.
        """
        with self._lock:
            self._maybe_reset_window()
            now = time.time()
            elapsed_hours = (now - self._window_start) / 3600.0
            if elapsed_hours <= 0:
                return 0.0

            current = self._user_spend.get(user, 0.0) if user else self._current_spend

            rate = current / elapsed_hours
            return rate * hours

    def is_valid_user(self, user: str) -> bool:
        """Check if a user has a budget set and hasn't exceeded it.

        Returns ``True`` if the user has a budget and their current spend
        is within that budget, ``False`` otherwise.
        """
        with self._lock:
            self._maybe_reset_window()
            if user not in self._user_budgets:
                return False
            user_spend = self._user_spend.get(user, 0.0)
            return user_spend <= self._user_budgets[user]
